src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+    case Serialize::TailStrategy::TailStrategy_RoundUp:
+        return TailStrategy::RoundUp;
+    case Serialize::TailStrategy::TailStrategy_GuardWithIf:
+        return TailStrategy::GuardWithIf;
+    case Serialize::TailStrategy::TailStrategy_Predicate:
+        return TailStrategy::Predicate;
+    case Serialize::TailStrategy::TailStrategy_PredicateLoads:
+        return TailStrategy::PredicateLoads;
+    case Serialize::TailStrategy::TailStrategy_PredicateStores:
+        return TailStrategy::PredicateStores;
+    case Serialize::TailStrategy::TailStrategy_ShiftInwards:
+        return TailStrategy::ShiftInwards;
+    case Serialize::TailStrategy::TailStrategy_Auto:
+        return TailStrategy::Auto;
+    default:
+        user_assert(false) << "unknown tail strategy " << tail_strategy << "\n";
+        return TailStrategy::RoundUp;
+    }
+}
+
+Split::SplitType Deserializer::deserialize_split_type(const Serialize::SplitType split_type) {
+    switch (split_type) {
+    case Serialize::SplitType::SplitType_SplitVar:
+        return Split::SplitType::SplitVar;
+    case Serialize::SplitType::SplitType_RenameVar:
+        return Split::SplitType::RenameVar;
+    case Serialize::SplitType::SplitType_FuseVars:
+        return Split::SplitType::FuseVars;
+    case Serialize::SplitType::SplitType_PurifyRVar:
+        return Split::SplitType::PurifyRVar;
+    default:
+        user_assert(false) << "unknown split type " << split_type << "\n";
+        return Split::SplitType::SplitVar;
+    }
+}
+
+DimType Deserializer::deserialize_dim_type(const Serialize::DimType dim_type) {
+    switch (dim_type) {
+    case Serialize::DimType::DimType_PureVar:
+        return DimType::PureVar;
+    case Serialize::DimType::DimType_PureRVar:
+        return DimType::PureRVar;
+    case Serialize::DimType::DimType_ImpureRVar:
+        return DimType::ImpureRVar;
+    default:
+        user_assert(false) << "unknown dim type " << dim_type << "\n";
+        return DimType::PureVar;
+    }
+}
+
+LoopAlignStrategy Deserializer::deserialize_loop_align_strategy(const Serialize::LoopAlignStrategy loop_align_strategy) {
+    switch (loop_align_strategy) {
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_AlignStart:
+        return LoopAlignStrategy::AlignStart;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_AlignEnd:
+        return LoopAlignStrategy::AlignEnd;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_NoAlign:
+        return LoopAlignStrategy::NoAlign;
+    case Serialize::LoopAlignStrategy::LoopAlignStrategy_Auto:
+        return LoopAlignStrategy::Auto;
+    default:
+        user_assert(false) << "unknown loop align strategy " << loop_align_strategy << "\n";
+        return LoopAlignStrategy::AlignStart;
+    }
+}
+
+ExternFuncArgument::ArgType Deserializer::deserialize_extern_func_argument_type(const Serialize::ExternFuncArgumentType extern_func_argument_type) {
+    switch (extern_func_argument_type) {
```

Comment:
Doing all of these via switch-with-default is probably the right way to do this, but the premature optimizer in my brain wants to set up some data-driven templated table nonsense. I'm glad you resisted that urge :-)